### PR TITLE
refactor(error): migrate function errors to anyhow

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

SQL functions now return `anyhow` error. Making it easy for implementors to construct errors using `anyhow!`, `bail!` and `.context()`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

### CODE REVIEW

**Comment by TennyZhuang** [comment]
Anyway, I like the PR so much. Just some concern about performance.

**Comment by wangrunji0408** [comment]
Any ideas on how to disable backtrace inside the `expr_impl` crate? 🥹

**Comment by TennyZhuang** [comment]
> Any ideas on how to disable backtrace inside the `expr_impl` crate? 🥹

@BugenZhao Can you give some insight?

A simple solution is `Box<dyn Error>` with some syntax sugur (e.g. context).

File: src/expr/macro/src/gen.rs [hunk 1]
```diff
@@ -321,13 +321,14 @@ impl FunctionAttr {
...
+        let error_context = format!("in function \"{}\"", self.name);
         output = match user_fn.return_type_kind {
             // XXX: we don't support void type yet. return null::int for now.
             _ if self.ret == "void" => quote! { { #output; Option::<i32>::None } },
             ReturnTypeKind::T => quote! { Some(#output) },
             ReturnTypeKind::Option => output,
-            ReturnTypeKind::Result => quote! { Some(#output?) },
-            ReturnTypeKind::ResultOption => quote! { #output? },
+            ReturnTypeKind::Result => quote! { Some(#output.context(#error_context)?) },
+            ReturnTypeKind::ResultOption => quote! { #output.context(#error_context)? },
```
Comment (xxchan) [inline src/expr/macro/src/gen.rs:331]: qq: So the most important change is here right? Could you teach me why `anyhow` is necessary for this?

Comment (wangrunji0408) [reply src/expr/macro/src/gen.rs:331]: I don't think this is the most important part. It is just a side-effect that `anyhow` makes it easy to add context around the error. 🤡

Comment (xxchan) [reply src/expr/macro/src/gen.rs:331]: Let me rephrase my points: I mean the most important change is the ability to `.context`, right? But does this require `anyhow`? 🤔 (I mean `anyhow` is the implementation detail, instead of the goal) I remember bugen's `thiserror_ext` also have context.

BTW, the most important user-facing change looks indeed like the function name. 🤣 Can you show me some other exciting examples?

Comment (xxchan) [reply src/expr/macro/src/gen.rs:331]: > I roughly remember bugen's thiserror_ext also have context.

Well, it seems to be `ContextInto`, instead of `ContextInfo` 🤡

Comment (wangrunji0408) [reply src/expr/macro/src/gen.rs:331]: From developer's view, they can freely use `anyhow` error and no longer need to consider which `ExprError` to use. 🤡

Comment (xxchan) [reply src/expr/macro/src/gen.rs:331]: Hmmm, I'm thinking about sth like this might suffice

```rust
struct ContextError {
    contexts: Vec<String>,
}

impl ContextError {
    fn context(self, context: impl Display) -> Self {
        let mut contexts = self.contexts;
        contexts.push(context.to_string());
        Self { contexts }
    }
}

// This trait refers to anyhow
trait Context<T> {
    /// Wrap the error value with additional context.
    fn context<C>(self, context: C) -> Result<T, ContextError>
    where
        C: Display + Send + Sync + 'static;

    /// Wrap the error value with additional context that is evaluated lazily
    /// only once an error does occur.
    fn with_context<C, F>(self, f: F) -> Result<T, ContextError>
    where
        C: Display + Send + Sync + 'static,
        F: FnOnce() -> C;
}

impl<T> Context<T> for Result<T, ContextError> {
    fn context<C>(self, context: C) -> Result<T, ContextError>
    where
        C: Display + Send + Sync + 'static,
    {
        // Not using map_err to save 2 useless frames off the captured backtrace
        // in ext_context.
        match self {
            Ok(ok) => Ok(ok),
            Err(error) => Err(error.context(context)),
        }
    }

    fn with_context<C, F>(self, context: F) -> Result<T, ContextError>
    where
        C: Display + Send + Sync + 'static,
        F: FnOnce() -> C,
    {
        match self {
            Ok(ok) => Ok(ok),
            Err(error) => Err(error.context(context())),
        }
    }
}

fn foo() -> Result<(), ContextError> {
    Err(ContextError {
        contexts: vec!["foo".to_string()],
    })
}

fn bar() -> Result<(), ContextError> {
    foo().context("wtf").context("woo")
}
```

File: src/expr/impl/src/scalar/cast.rs [hunk 2]
```diff
@@ -40,32 +40,37 @@ use thiserror_ext::AsReport;
 pub fn str_parse<T>(elem: &str, ctx: &Context) -> Result<T>
 where
     T: FromStr,
-    <T as FromStr>::Err: std::fmt::Display,
 {
-    elem.trim().parse().map_err(|err: <T as FromStr>::Err| {
-        ExprError::Parse(format!("{} {}", ctx.return_type, err).into())
+    elem.trim().parse().ok().with_context(|| {
```
Comment (xxchan) [inline src/expr/impl/src/scalar/cast.rs:44]: `.ok()` is not necessary here? 🤔

File: src/expr/macro/src/gen.rs (hunk 1 above)
Comment (xxchan) [reply src/expr/macro/src/gen.rs:331]: Oh, this is just similar to tygg's comment above

> A simple solution is Box<dyn Error> with some syntax sugur (e.g. context).

**Comment by BugenZhao** [comment]
> > Any ideas on how to disable backtrace inside the `expr_impl` crate? 🥹
> 
> @BugenZhao Can you give some insight?

I guess there's no way to disable that. The backtrace will always be captured in nightly channel.

https://docs.rs/anyhow/latest/src/anyhow/lib.rs.html#131-132

File: src/expr/macro/src/gen.rs (hunk 1 above)
Comment (BugenZhao) [reply src/expr/macro/src/gen.rs:331]: FYI, the implementation of `anyhow::Context::context` uses an anonymous error type to provide the context and put the original error as its `source`, in which way the source chain is correctly maintained.

https://docs.rs/anyhow/latest/src/anyhow/error.rs.html#305-308

Comment (xxchan) [reply src/expr/macro/src/gen.rs:331]: Actually I view these expr errors as leaf errors. So just feel some string stuff is enough.

**Comment by xxchan** [comment]
We can use a forked `anyhow` without backtraces if really necessary. 🤪 I think no much effort is needed.

File: src/expr/impl/src/scalar/cast.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/expr/impl/src/scalar/cast.rs:44]: If removed, a complex type constraint would be required.
```
error[E0277]: `<T as FromStr>::Err` cannot be sent between threads safely
  --> src/expr/impl/src/scalar/cast.rs:44:25
   |
44 |     elem.trim().parse().with_context(|| {
   |                         ^^^^^^^^^^^^ `<T as FromStr>::Err` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `<T as FromStr>::Err`
   = note: required for `std::result::Result<T, <T as FromStr>::Err>` to implement `anyhow::Context<T, <T as FromStr>::Err>`
help: consider further restricting the associated type
   |
42 |     T: FromStr, <T as FromStr>::Err: std::marker::Send
   |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by xxchan** [comment]
about the backtrace's overhead:  https://github.com/dtolnay/anyhow/issues/41#issuecomment-578395725

and some numbers I tried previously https://github.com/risingwavelabs/risingwave/issues/6205#issuecomment-1502015073

I think the most important question is whether this sentence holds true 🤔

> Errors in Expr often occur in succession

**Comment by TennyZhuang** [comment]
I have a rough idea (partially inspired by @yuhao-su).

> If using Rust ≥ 1.65, a backtrace is captured and printed with the error if the underlying error type does not already provide its own. In order to see backtraces, they must be enabled through the environment variables described in [std::backtrace](https://doc.rust-lang.org/std/backtrace/index.html#environment-variables):

According to the doc of anyhow, it seems that they will uses [`error_generic_member_access`](https://github.com/rust-lang/rust/issues/99301) feature to acquire the backtrace from its underlying Error. Which means that we can provide a fake `Backtrace` to misguide anyhow, and then anyhow may giveup to capture a backtrace by themselves.

```rust
/// Underlying, real error in the implementation
struct ParseIntError;

struct DisableBacktrace<E: Error>(E);

impl Error for DisabeBacktrace {
    fn provide<'a>(&'a self, req: &mut Demand<'a>) {
        // Pseudo code, need more investigation
        req.provide(Backtrace::disabled);
    }
}

anyhow::Error::new(DisableBacktrace(ParseIntError));
```

**Comment by wangrunji0408** [comment]
After discussion with @TennyZhuang, we decided to encourage function implementors to return custom error types rather than anyhow, so that functions can be reused outside of expression framework. The `#[function]` macro will accept any error type and wrap them in an anyhow error with necessary context.

**Comment by BugenZhao** [comment]
> Which means that we can provide a fake `Backtrace` to misguide anyhow, and then anyhow may giveup to capture a backtrace by themselves.

If we really want a backtrace in the upper layer, it's worth noting that we might be misguided as well. A simple workaround could be not marking `#[backtrace]` to the `ExprError` type.

**Comment by BugenZhao** [comment]
> It is just a side-effect that anyhow makes it easy to add context around the error. 🤡

If this is the main motivation, I guess we can achieve similar effects with `thiserror_ext` (inspired by discussion w/ @xiangjinwu). 😄

Check this example: https://github.com/BugenZhao/thiserror-ext/blob/8e8d7c65f3ddbc1f9097e9200a83f8d3bf9d2a70/examples/context.rs

**Comment by xxchan** [comment]
> Check this example: [BugenZhao/thiserror-ext@`8e8d7c6`/examples/context.rs](https://github.com/BugenZhao/thiserror-ext/blob/8e8d7c65f3ddbc1f9097e9200a83f8d3bf9d2a70/examples/context.rs)

Previously I was thinking adding context to the outer struct. 🤣 This solution is quite simple (reminds me of linked list)
